tests/controller_tests/test_historical_ingest.py

```diff
@@ -0,0 +1,439 @@
+        submission_date=datetime(2023, 5, 1),
+        reporting_period_start=datetime(2023, 4, 1),
+        reporting_period_end=datetime(2023, 4, 1),
+        reporting_round=3,
+    )
+    db.session.add(sub)
+    read_sub = Submission.query.first()
+    organisation = Organisation(
+        organisation_name="Romulan Star Empire",
+        geography="Romulas",
+    )
+    db.session.add(organisation)
+    read_org = Organisation.query.first()
+    prog = Programme(
+        programme_id="TD-ROM",
+        programme_name="Tal Shiar",
+        fund_type_id="TD",
+        organisation_id=read_org.id,
+    )
+    db.session.add(prog)
+    read_prog = Programme.query.first()
+    proj = Project(
+        project_id="TD-ROM-01",
+        submission_id=read_sub.id,
+        programme_id=read_prog.id,
+        project_name="Patrol the neutral zone",
+        primary_intervention_theme="N/A",
+        location_multiplicity="Single",
+        locations="ROM EMP",
+    )
+    db.session.add(proj)
+
+    populate_db_historical_data(r2_workbook_mockup, INGEST_MAPPINGS)
+
+    # check R3 data is still present
+    org = Organisation.query.filter_by(organisation_name="Romulan Star Empire").first()
+    assert org is not None
+
+    prog = Programme.query.filter_by(programme_id="TD-ROM").first()
+    assert prog is not None
+
+    child = Project.query.filter_by(project_id="TD-ROM-01").first()
+    assert child is not None
+
+    populate_db_historical_data(r1_workbook_mockup, INGEST_MAPPINGS)
+
+    # only expect one result as existing org should not be updated
+    org = Organisation.query.filter_by(organisation_name="Romulan Star Empire").all()
+    assert len(org) == 1
+    assert org[0] == read_org
+
+    # expect R3 programme to not be overwritten
+    prog = Programme.query.filter_by(programme_id="TD-ROM").all()
+    assert len(prog) == 1
+    assert prog[0] == read_prog
+
+    # expect both projects for the programme have been ingested
+    proj1 = Project.query.filter_by(project_id="TD-ROM-01").first()
+    proj2 = Project.query.filter_by(project_id="TD-ROM-02").first()
+    assert proj1 is not None and proj2 is not None
+
+
+def test_historical_populate_db_round_one_then_two(test_client, r2_workbook_mockup, r1_workbook_mockup):
+    """Tests that round one can be ingested, and then round two, updating conflicting programmes.
+
+    Test also ensures that Round 1 children are not effected by parental changes.
+    """
+    populate_db_historical_data(r1_workbook_mockup, INGEST_MAPPINGS)
+    populate_db_historical_data(r2_workbook_mockup, INGEST_MAPPINGS)
+
+    # check programme updated (should have new name)
+    prog = Programme.query.filter_by(programme_id="TD-FED").with_entities(Programme.programme_name).all()
+    assert len(prog) == 1
+    assert ("Rebuild Bajor",) in prog
+
+    # ensure child with same programme not deleted or overwritten
+    child = RiskRegister.query.filter_by(risk_name="Dominion Attack").first()
```

Comment:
this test assertion would be more concise if it actually filtered the RiskRegister entity by filtering on the programme. As currently there is a lot of assumption implied by the comment.

Likewise the part above "# check programme updated (should have new name)" could do a more explicit test, ie, run the query before the R2 upsert and then after, and show they are not the same.

The risk with not doing that (ie just saying it is the same before the upsert in a comment is that any subsequent person editing this test have to trust that comment, but if it's tested that it's always explicitly demonstrated and tested.

Reply:
For instance, in this current test, in order to check the details of exactly what is being tested for, I'd have to go searching all through the fixture setup code to work out what  the comment was referring to (in this case, that the child actually existed before we inserted R2, and go looking to find out what it was, and that it didn't match the one in the assertion here)